app/src/main/java/com/lightricks/feedexercise/util/extensions.kt

```diff
@@ -0,0 +1,37 @@
+package com.lightricks.feedexercise.util
+
+import com.lightricks.feedexercise.data.FeedItem
+import com.lightricks.feedexercise.database.FeedItemEntity
+import com.lightricks.feedexercise.network.FeedItemResponse
+
+//taken from here: https://medium.com/@BladeCoder/kotlin-singletons-with-argument-194ef06edd9e
+open class SingletonHolder<out T: Any, in A>(creator: (A) -> T) {
```

Comment:
This looks correct, although it's verbose. Probably, adapted from Java.
In Kotlin you can make it a little simpler.
For example in FeedDatabase.kt you could do this:
```
private lateinit var INSTANCE: FeedDatabase

fun getDatabase(context: Context): FeedDatabase {
    synchronized(FeedDatabase::class.java) {
        if (!::INSTANCE.isInitialized) {
            INSTANCE = Room.databaseBuilder(context.applicationContext,
                FeedDatabase::class.java, "feed_database")
                .build()
        }
    }

    return INSTANCE
}
```

Reply:
To my humble opinion, it has nothing to do with java, it just a generic extensible way to encapsulate the Singleton with parameters creation, in a neat and clean way.

Reply:
I agree, that it's generic :)